sklearn/preprocessing/_encoders.py

```diff
@@ -0,0 +1,845 @@
+        (in order corresponding with output of ``transform``).
+
+    active_features_ : array
+        Indices for active features, meaning values that actually occur
+        in the training set. Only available when n_values is ``'auto'``.
+
+        .. deprecated:: 0.20
+            The `active_features_` attribute was deprecated in version
+            0.20 and will be removed in 0.22.
+
+    feature_indices_ : array of shape (n_features,)
+        Indices to feature ranges.
+        Feature ``i`` in the original data is mapped to features
+        from ``feature_indices_[i]`` to ``feature_indices_[i+1]``
+        (and then potentially masked by `active_features_` afterwards)
+
+        .. deprecated:: 0.20
+            The `feature_indices_` attribute was deprecated in version
+            0.20 and will be removed in 0.22.
+
+    n_values_ : array of shape (n_features,)
+        Maximum number of values per feature.
+
+        .. deprecated:: 0.20
+            The `n_values_` attribute was deprecated in version
+            0.20 and will be removed in 0.22.
+
+    Examples
+    --------
+    Given a dataset with two features, we let the encoder find the unique
+    values per feature and transform the data to a binary one-hot encoding.
+
+    >>> from sklearn.preprocessing import OneHotEncoder
+    >>> enc = OneHotEncoder(handle_unknown='ignore')
+    >>> X = [['Male', 1], ['Female', 3], ['Female', 2]]
+    >>> enc.fit(X)
+    ... # doctest: +ELLIPSIS
+    OneHotEncoder(categorical_features='all', categories='auto',
+           dtype=<... 'numpy.float64'>, handle_unknown='ignore',
+           n_values=None, sparse=True)
+
+    >>> enc.categories_
+    [array(['Female', 'Male'], dtype=object), array([1, 2, 3], dtype=object)]
+    >>> enc.transform([['Female', 1], ['Male', 4]]).toarray()
+    array([[1., 0., 1., 0., 0.],
+           [0., 1., 0., 0., 0.]])
+    >>> enc.inverse_transform([[0, 1, 1, 0, 0], [0, 0, 0, 1, 0]])
+    array([['Male', 1],
+           [None, 2]], dtype=object)
+
+    See also
+    --------
+    sklearn.preprocessing.OrdinalEncoder : performs an ordinal (integer)
+      encoding of the categorical features.
+    sklearn.feature_extraction.DictVectorizer : performs a one-hot encoding of
+      dictionary items (also handles string-valued features).
+    sklearn.feature_extraction.FeatureHasher : performs an approximate one-hot
+      encoding of dictionary items or strings.
+    sklearn.preprocessing.LabelBinarizer : binarizes labels in a one-vs-all
+      fashion.
+    sklearn.preprocessing.MultiLabelBinarizer : transforms between iterable of
+      iterables and a multilabel format, e.g. a (samples x classes) binary
+      matrix indicating the presence of a class label.
+    """
+
+    def __init__(self, n_values=None, categorical_features=None,
+                 categories=None, sparse=True, dtype=np.float64,
+                 handle_unknown='error'):
+        self._categories = self.categories = categories
```

Comment:
I removed the deprecations of the params, and any validation in `fit`, but in the end I did keep the "underscored" versions of `categories`, `n_values` and `categorical_features`, to be able to know what the user passed itself, and thus raise the appropriate warning. 

The alternative is that throughout the actual implementation I have to add a lot of checks whether eg `self.categories is None` in addition to `self.categories == 'auto'` because None as default actually means 'auto' (and the same for the other two). 
But personally, I find that that makes the implementation much more complex compared to keeping those underscored versions of the initial passed values by the user (and setting the non-underscored versions to the appropriate default value in `_handle_deprecations` (where all that logic is then centralized).

Reply:
But your underscores break set_params, which is unacceptable.

You can create all the underscore-prefixed attribs you like in fit to store validated versions and simplify downstream code. Your init must only set attributes by their arg name.

Reply:
Whoops, yes, that's a good catch. Indeed, I can create the underscored versions in `fit` and switch the purpose of the underscored vs non-underscored ones.

The problem I was trying to solve with keeping track of both, is that I wanted to warn on each fit, not only the first (because I translate certain values for `n_values` to `categories`, the second time you call `fit`, `categories` is set by the code, and no warning will be raised). But actually, maybe that's fine (or even better?) to only warn the first time?